lib/Graph/Log.cpp

```diff
@@ -0,0 +1,149 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Graph/Log.h"
+#include "glow/Graph/Node.h"
+
+namespace glow {
+
+llvm::cl::opt<bool>
+    dumpCompilationLogOpt("dump-compilation-log",
+                          llvm::cl::desc("Dump compilation log"));
+
+void LogContext::addLogContent(llvm::StringRef logContent) {
+  logContents_.push_back(logContent);
+}
+
+void LogContext::addLogScope(llvm::StringRef scopeName) {
+  logScopes_.push_back(scopeName);
+  currentFullScope_ = "";
+  for (auto s : logScopes_) {
+    currentFullScope_ += "->";
+    currentFullScope_ += s;
+  }
+}
+
+void LogContext::popBackLogScope() {
+  if (logScopes_.size() > 0) {
+    logScopes_.pop_back();
+  }
+  currentFullScope_ = "";
+  for (auto s : logScopes_) {
+    currentFullScope_ += "->";
+    currentFullScope_ += s;
+  }
+}
+void LogContext::dumpLog(llvm::StringRef funcName) {
+  if (dumpCompilationLogOpt) {
+    std::string compileLogFilename = funcName;
+    compileLogFilename.append("_compile.log");
+
+    llvm::outs() << "Writing compilation log file for Module to: "
+                 << compileLogFilename << '\n';
+    std::error_code EC;
+    llvm::raw_fd_ostream myfile(compileLogFilename, EC);
+    for (auto s : logContents_) {
+      myfile << s;
```

Comment:
what happens if file already exist?

Reply:
it will be overwritten. But I guess we don't really care if there was a previous log file there. What we care is the content that is being written into the file right now.

Reply:
It should also be the user's job to archive previous log files into some other directory.

Reply:
But i guess i can also have another command line option to pass the specific name for the log file in the future.